fluss-client/src/main/java/com/alibaba/fluss/client/write/DynamicWriteBatchSizeEstimator.java

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright (c) 2025 Alibaba Group Holding Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.alibaba.fluss.client.write;
+
+import com.alibaba.fluss.annotation.Internal;
+import com.alibaba.fluss.annotation.VisibleForTesting;
+import com.alibaba.fluss.metadata.PhysicalTablePath;
+import com.alibaba.fluss.utils.concurrent.FlussScheduler;
+import com.alibaba.fluss.utils.concurrent.Scheduler;
+import com.alibaba.fluss.utils.types.Tuple2;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import static com.alibaba.fluss.utils.concurrent.LockUtils.inLock;
+
+/** An estimator to estimate the buffer usage of a writeBatch. */
+@Internal
+@ThreadSafe
+public class DynamicWriteBatchSizeEstimator {
+
+    private static final Logger LOG = LoggerFactory.getLogger(DynamicWriteBatchSizeEstimator.class);
+
+    private static final int MAX_QUEUE_SIZE = 100;
+    private static final double DIFF_RATIO_THRESHOLD = 0.2;
+    static final double INCREASE_RATIO = 0.2;
+    private final Lock lock = new ReentrantLock();
+    private final int initialBatchSize;
+    private final int bufferTotalSize;
+    private final int pageSize;
+
+    @GuardedBy("lock")
+    private final Map<PhysicalTablePath, WriteBatchSizeStatus> writeBatchSizeStatusMap;
+
+    public DynamicWriteBatchSizeEstimator(
+            int initialBatchSize, int bufferTotalSize, int pageSize, long estimateInterval) {
+        this.writeBatchSizeStatusMap = new HashMap<>();
+        this.initialBatchSize = initialBatchSize;
+        this.bufferTotalSize = bufferTotalSize;
+        this.pageSize = pageSize;
+
+        Scheduler scheduler = new FlussScheduler(1);
```

Comment:
We don't need a thread to do this work. Just dynamic configuring the new batch size for each closed batch.